# Тест по ML
Выполнил Давыдов Данил
davydov-danil@bk.ru

In [1]:
# импортируем библиотеки
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import classification_report
from sklearn.metrics import roc_auc_score
from xgboost import XGBClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold

In [2]:
# загружаем данные
data = pd.read_excel('result_set ML тест.xlsx')

# Анализируем датасет

In [3]:
data.head()

Ft 0      Ft 1      Ft 2      Ft 3       Ft 4      Ft 5       Ft 6  \
0  1.620967  1.056248  0.114203  6.175593   5.625789  0.581535   7.360085   
1  1.274494  0.814288  0.727210  3.606359   1.493780  7.810750  20.314335   
2  0.569931  1.607170  0.479581  3.521474   1.663726  2.297903  12.839282   
3  0.547560  1.724393  0.682358  1.809472  10.149721  2.999319  49.108530   
4  1.798884  1.748245  1.045646  1.560427   1.269053  1.403952   5.270951   

   Ft 7    Ft 8      Ft 9  Ft 10  Ft 11  Ft 12  Ft 13  Ft14        Ft15  \
0    68    male  very_low  535.0   True      5      5     0         NaN   
1    24  female      high  846.0  False      3      1     1  quite rich   
2    67    male      high  782.0  False      0      0     0    moderate   
3    42    male    medium  132.0  False      3      3     0  quite rich   
4    84    male  very_low  708.0  False      0      0     0         NaN   

                  Ft16  Ft17  mark  
0                  car     0     0  
1                  car     1     0  
2  furniture/equipment     1     0  
3                  car     0     0  
4             radio/TV     1     0

Датасет содержит 18 фичей и 1 бинарный таргет

In [4]:
data.describe()

Ft 0         Ft 1         Ft 2         Ft 3         Ft 4  \
count  5000.000000  5000.000000  5000.000000  5000.000000  5000.000000   
mean      1.157080     1.060064     0.839764     1.439738     2.081017   
std       0.702209     0.420866     0.508024     1.402939     3.022327   
min       0.066066     0.177825     0.017505     0.001728     0.000007   
25%       0.660346     0.765661     0.498008     0.484095     0.407973   
50%       1.000119     0.992588     0.731577     1.032661     1.253750   
75%       1.471052     1.292738     1.052030     1.921633     2.681746   
max       5.942049     3.331410     6.169126    19.940927    86.940693   

              Ft 5         Ft 6         Ft 7        Ft 10        Ft 12  \
count  5000.000000  5000.000000  5000.000000  4936.000000  5000.000000   
mean      1.665331    14.005111    51.103600   552.702998     2.712000   
std       1.716978    14.416559    19.526688   249.698601    20.152784   
min       0.002628     0.190074    18.000000    99.000000 -1000.000000   
25%       0.468759     4.817639    34.000000   305.000000     1.000000   
50%       1.202163     9.584938    51.000000   560.000000     4.000000   
75%       2.291755    17.884976    68.000000   784.000000     5.000000   
max      23.327828   157.312632    85.000000   981.000000     5.000000   

             Ft 13         Ft14         Ft17         mark  
count  5000.000000  5000.000000  5000.000000  5000.000000  
mean      1.873000     0.172200     0.615400     0.012800  
std      20.084311     0.377592     0.486549     0.112422  
min   -1000.000000     0.000000     0.000000     0.000000  
25%       2.000000     0.000000     0.000000     0.000000  
50%       3.000000     0.000000     1.000000     0.000000  
75%       3.000000     0.000000     1.000000     0.000000  
max       5.000000     1.000000     1.000000     1.000000

Датасет состоит из 5000 записей

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 19 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Ft 0    5000 non-null   float64
 1   Ft 1    5000 non-null   float64
 2   Ft 2    5000 non-null   float64
 3   Ft 3    5000 non-null   float64
 4   Ft 4    5000 non-null   float64
 5   Ft 5    5000 non-null   float64
 6   Ft 6    5000 non-null   float64
 7   Ft 7    5000 non-null   int64  
 8   Ft 8    5000 non-null   object 
 9   Ft 9    5000 non-null   object 
 10  Ft 10   4936 non-null   float64
 11  Ft 11   4950 non-null   object 
 12  Ft 12   5000 non-null   int64  
 13  Ft 13   5000 non-null   int64  
 14  Ft14    5000 non-null   int64  
 15  Ft15    4183 non-null   object 
 16  Ft16    5000 non-null   object 
 17  Ft17    5000 non-null   int64  
 18  mark    5000 non-null   int64  
dtypes: float64(8), int64(6), object(5)
memory usage: 742.3+ KB


В датасете есть пропущенные данные как в категориальных фичах, так и в некатегориальных.

In [6]:
data.isnull().sum(axis=0)

Ft 0       0
Ft 1       0
Ft 2       0
Ft 3       0
Ft 4       0
Ft 5       0
Ft 6       0
Ft 7       0
Ft 8       0
Ft 9       0
Ft 10     64
Ft 11     50
Ft 12      0
Ft 13      0
Ft14       0
Ft15     817
Ft16       0
Ft17       0
mark       0
dtype: int64

Количество пропущенных значений не слишком велико, такие значения можно заменить на значание из той же колонки используя стратегию среднего или стратегию моды(наиболее часто встречающееся).

In [7]:
print(data['mark'].value_counts())

mark
0    4936
1      64
Name: count, dtype: int64


Датасет не сбалансирован по таргету, положительный класс находится в меньшинстве. Положительный класс занимает 1.28% от всех записей, то есть датасет очень сильно несбалансирован.

In [8]:
object_cols = [column for column in data.columns if data[column].dtype=='object']

In [9]:
object_cols

['Ft 8', 'Ft 9', 'Ft 11', 'Ft15', 'Ft16']

Все категорильные фичи

In [10]:
for column in object_cols:
    print(data[column].value_counts())

Ft 8
male      4124
female     876
Name: count, dtype: int64
Ft 9
high        2482
medium      1545
very_low     520
low          453
Name: count, dtype: int64
Ft 11
True     3425
False    1525
Name: count, dtype: int64
Ft15
little        1505
moderate      1171
quite rich     784
rich           723
Name: count, dtype: int64
Ft16
radio/TV               1138
car                     965
furniture/equipment     628
education               492
repairs                 486
vacation/others         478
business                460
domestic appliances     353
Name: count, dtype: int64


Количество уникальных значений в категорильных колонках не слишком велико, в будущем можно использовать onehot encoding для кодирования категорильных фич.

# Предобработка

In [11]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline


# делим данные на обучающую и тестовую выборки
X, y = data.iloc[:, :-1], data.iloc[:, -1]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, stratify=y, random_state=31, shuffle=True)

In [12]:
# определяем некатегориальные фичи
numeric_cols = [column for column in data.columns if column not in object_cols and column!=data.columns[-1]]

# imputer для вставки пропущенных значений, среднее для некатегориальных и мода для категорильных
imputer = ColumnTransformer([
    ('numeric_imputer', SimpleImputer(strategy='mean'), numeric_cols),
    ('categorical_imputer', SimpleImputer(strategy='most_frequent'), object_cols)
    ],
)

# scaler для некатегориальных переменных и encoder для категориальных, используя onehot encoder
scaler_encoder = ColumnTransformer([
    ('numeric_scaler', StandardScaler(), list(range(len(numeric_cols)))),
    ('categorical_encoder', OneHotEncoder(sparse_output=False), list(range(len(numeric_cols), len(numeric_cols) + len(object_cols))))
    ],
)

# делаем вывод трансформеров в виде pandas.DataFrame
imputer.set_output(transform='pandas')
scaler_encoder.set_output(transform='pandas')

# создаем pipeline для предобработки данных
transformer = Pipeline([
    ('imputing', imputer),
    ('scaling_encoding', scaler_encoder),]
)
transformer.set_output(transform='pandas')

Pipeline(steps=[('imputing',
                 ColumnTransformer(transformers=[('numeric_imputer',
                                                  SimpleImputer(),
                                                  ['Ft 0', 'Ft 1', 'Ft 2',
                                                   'Ft 3', 'Ft 4', 'Ft 5',
                                                   'Ft 6', 'Ft 7', 'Ft 10',
                                                   'Ft 12', 'Ft 13', 'Ft14',
                                                   'Ft17']),
                                                 ('categorical_imputer',
                                                  SimpleImputer(strategy='most_frequent'),
                                                  ['Ft 8', 'Ft 9', 'Ft 11',
                                                   'Ft15', 'Ft16'])])),
                ('scaling_encoding',
                 ColumnTransformer(transformers=[('numeric_scaler',
                                                  StandardScaler(),
                                                  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9,
                                                   10, 11, 12]),
                                                 ('categorical_encoder',
                                                  OneHotEncoder(sparse_output=False),
                                                  [13, 14, 15, 16, 17])]))])

In [13]:
# преобразуем данные
X_train_transformed = transformer.fit_transform(X_train)
X_test_transformed = transformer.transform(X_test)

In [14]:
X_train_transformed.shape

(4000, 33)

# Обучение модели

Наша проблема стоит в бинарной классификации, первая идея - попробовать логистическую регрессию.   
Так как таргет несбалансирован есть смысл применить технику undersampling или/KFold кроссвалидацию, а также указать веса классов для наказания неверных предсказаний класса-меньшинства.  
Наиболее подходящая технология XGBoost в силу своей быстроты и эффективности, но я попробую и модели из sklearn: LogisticRegression, RandomForestClassifier.  
Выбор метрики очень важен, так recall очень важен если нужно поймать как можно больше положительных классов правильно, roc-auc score же определяет насколько модель умеет разделять классы.  
Так как основное напрявление вакансии fraud detection, то важно чтобы модель ловила как можно больше положительных классов(True Positive) и при этом как можно меньше помечала отрицательные классы как положительные(False Positive).  
Поэтому основная метрика - это roc-auc score.

In [15]:
# пробуем логистическую регрессию, перебираем параметры gridsearh
lr_params = {
    'C': [0.5, 1, 2, 5],
    'class_weight': [{0:0.013, 1:1}, {0:1, 1:1}]
}
grid_lr = GridSearchCV(LogisticRegression(), param_grid=lr_params, verbose=2, n_jobs=-1, scoring='roc_auc')
grid_lr.fit(X_train_transformed, y_train)

Fitting 5 folds for each of 8 candidates, totalling 40 fits
[CV] END ...............C=0.5, class_weight={0: 0.013, 1: 1}; total time=   0.0s
[CV] END ...................C=0.5, class_weight={0: 1, 1: 1}; total time=   0.0s
[CV] END ...................C=0.5, class_weight={0: 1, 1: 1}; total time=   0.0s
[CV] END .................C=1, class_weight={0: 0.013, 1: 1}; total time=   0.0s
[CV] END ...............C=0.5, class_weight={0: 0.013, 1: 1}; total time=   0.0s
[CV] END .................C=1, class_weight={0: 0.013, 1: 1}; total time=   0.0s
[CV] END .................C=1, class_weight={0: 0.013, 1: 1}; total time=   0.0s
[CV] END .................C=1, class_weight={0: 0.013, 1: 1}; total time=   0.0s
[CV] END .................C=1, class_weight={0: 0.013, 1: 1}; total time=   0.0s
[CV] END ...............C=0.5, class_weight={0: 0.013, 1: 1}; total time=   0.0s
[CV] END ...................C=0.5, class_weight={0: 1, 1: 1}; total time=   0.0s
[CV] END .....................C=1, class_weight={

GridSearchCV(estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': [0.5, 1, 2, 5],
                         'class_weight': [{0: 0.013, 1: 1}, {0: 1, 1: 1}]},
             scoring='roc_auc', verbose=2)

In [16]:
# ищем порог, который максимизирует roc-auc score
best_score, best_treshold = 0, np.nan
for t in np.linspace(0.1, 0.9, 20):
    y_pred = grid_lr.predict_proba(X_test_transformed)
    y_pred = y_pred[:, 1] > t
    if roc_auc_score(y_test, y_pred) > best_score:
        best_score, best_treshold = roc_auc_score(y_test, y_pred), t
    
y_pred = grid_lr.predict_proba(X_test_transformed)
y_pred = y_pred[:, 1] > best_treshold
print('Best treshold: ', best_treshold)
print(classification_report(y_test, y_pred))
print('Best roc-auc score: ', roc_auc_score(y_test, y_pred))

Best treshold:  0.31052631578947365
              precision    recall  f1-score   support

           0       1.00      0.67      0.80       987
           1       0.03      0.77      0.06        13

    accuracy                           0.68      1000
   macro avg       0.51      0.72      0.43      1000
weighted avg       0.98      0.68      0.79      1000

Best roc-auc score:  0.7214948172394982


Повторяем процедуру для случайного леса

In [17]:
rf_params = {
    'n_estimators': [100, 200, 300, 500],
    'class_weight': [{0:0.013, 1:1}, {0:1, 1:1}],
    'max_depth': [5, 10, 15],
    'ccp_alpha': [0.01, 0.025, 0.03]
}
grid_rf = GridSearchCV(RandomForestClassifier(), param_grid=rf_params, verbose=2, n_jobs=-1, scoring='roc_auc')
grid_rf.fit(X_train_transformed, y_train)

# ищем порог, который максимизирует roc-auc score
best_score, best_treshold = 0, np.nan
for t in np.linspace(0.1, 0.9, 20):
    y_pred = grid_rf.predict_proba(X_test_transformed)
    y_pred = y_pred[:, 1] > t
    if roc_auc_score(y_test, y_pred) > best_score:
        best_score, best_treshold = roc_auc_score(y_test, y_pred), t
    
y_pred = grid_rf.predict_proba(X_test_transformed)
y_pred = y_pred[:, 1] > best_treshold
print('Best treshold: ', best_treshold)
print(classification_report(y_test, y_pred))
print('Best roc-auc score: ', roc_auc_score(y_test, y_pred))

Fitting 5 folds for each of 72 candidates, totalling 360 fits
[CV] END ccp_alpha=0.01, class_weight={0: 0.013, 1: 1}, max_depth=5, n_estimators=100; total time=   0.4s
[CV] END ccp_alpha=0.01, class_weight={0: 0.013, 1: 1}, max_depth=5, n_estimators=100; total time=   0.4s
[CV] END ccp_alpha=0.01, class_weight={0: 0.013, 1: 1}, max_depth=5, n_estimators=100; total time=   0.4s
[CV] END ccp_alpha=0.01, class_weight={0: 0.013, 1: 1}, max_depth=5, n_estimators=100; total time=   0.4s
[CV] END ccp_alpha=0.01, class_weight={0: 0.013, 1: 1}, max_depth=5, n_estimators=100; total time=   0.4s
[CV] END ccp_alpha=0.01, class_weight={0: 0.013, 1: 1}, max_depth=5, n_estimators=200; total time=   0.8s
[CV] END ccp_alpha=0.01, class_weight={0: 0.013, 1: 1}, max_depth=5, n_estimators=200; total time=   0.8s
[CV] END ccp_alpha=0.01, class_weight={0: 0.013, 1: 1}, max_depth=5, n_estimators=200; total time=   0.8s
[CV] END ccp_alpha=0.01, class_weight={0: 0.013, 1: 1}, max_depth=5, n_estimators=200; tot

In [18]:
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline as imbPipeline

# пробуем XGBoost классификатор
weights = [1, 10, 25, 50, 75, 99, 100, 1000]
lrate = [0.1, 0.01, 0.2, 0.3]
max_depth = [5, 10, 20]
alpha = [5, 10, 15, 20]

# применяем KFold кроссвалидацию
param_grid = dict(scale_pos_weight=weights, learning_rate=lrate, max_depth=max_depth, alpha=alpha)
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
model = XGBClassifier()
# используем технику SMOTE, то есть oversampling
steps = [('over', SMOTE()), ('model', GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=-1, cv=cv, scoring='roc_auc'))]
pipeline = imbPipeline(steps=steps)
pipeline.fit(X_train_transformed, y_train)


Pipeline(steps=[('over', SMOTE()),
                ('model',
                 GridSearchCV(cv=RepeatedStratifiedKFold(n_repeats=3, n_splits=10, random_state=1),
                              estimator=XGBClassifier(base_score=None,
                                                      booster=None,
                                                      callbacks=None,
                                                      colsample_bylevel=None,
                                                      colsample_bynode=None,
                                                      colsample_bytree=None,
                                                      device=None,
                                                      early_stopping_rounds=None,
                                                      enable_categorical=False,
                                                      eval_metric=None,
                                                      feature_type...
                                                      max_depth=None,
                                                      max_leaves=None,
                                                      min_child_weight=None,
                                                      missing=nan,
                                                      monotone_constraints=None,
                                                      multi_strategy=None,
                                                      n_estimators=None,
                                                      n_jobs=None,
                                                      num_parallel_tree=None,
                                                      random_state=None, ...),
                              n_jobs=-1,
                              param_grid={'alpha': [5, 10, 15, 20],
                                          'learning_rate': [0.1, 0.01, 0.2,
                                                            0.3],
                                          'max_depth': [5, 10, 20],
                                          'scale_pos_weight': [1, 10, 25, 50,
                                                               75, 99, 100,
                                                               1000]},
                              scoring='roc_auc'))])

In [19]:
# ищем порог, который максимизирует roc-auc score
best_score, best_treshold = 0, np.nan
for t in np.linspace(0.01, 0.9, 50):
    y_pred = pipeline['model'].best_estimator_.predict_proba(X_test_transformed)
    y_pred = y_pred[:, 1] > t
    if roc_auc_score(y_test, y_pred) > best_score:
        best_score, best_treshold = roc_auc_score(y_test, y_pred), t
    
y_pred = pipeline['model'].best_estimator_.predict_proba(X_test_transformed)
y_pred = y_pred[:, 1] > best_treshold
print('Best treshold: ', best_treshold)
print(classification_report(y_test, y_pred))
print('Best roc-auc score: ', roc_auc_score(y_test, y_pred))

Best treshold:  0.028163265306122447
              precision    recall  f1-score   support

           0       1.00      0.89      0.94       987
           1       0.09      0.85      0.16        13

    accuracy                           0.89      1000
   macro avg       0.54      0.87      0.55      1000
weighted avg       0.99      0.89      0.93      1000

Best roc-auc score:  0.8668459200374095


Таким образом большинство положительных классов определено верно, а так как положительный класс является классом-меньшинством, то нам важно наиболее полно определить именно его, а объекты отрицательного класса, который определились как положительный, не сильно влияют на результат. 
Скорее всего, нужно определить является ли операцияп по счету мошеннической, и тогда нужно заблокировать действительно мошеннические переводы, а случайно заблокированные немошеннические можно разблокировать при дальнейшей свящи с клиентом.

# Итог

Я провел небольшой анализ датасета, выявил его особенности.  
Далее, предобработал данные, выявил основную метрику roc-auc score и обучил 3 модели.  
По итогу, лучше всех справляетя XGBoostClassifier с treshold-ом ~0.064. В таком случае основная метрика составляет ~0.869 и precision-recall-f1 метрики удовлетворительны.  
В fraud-detection данные обычно всегда несбалансированы и их очень много. Поэтому разные техники как undersampling, oversampling, crossvalidation очень нужны.